diff --git a/examples/gpuExamples/cavity3d/cavity3d.cpp b/examples/gpuExamples/cavity3d/cavity3d.cpp
index 6910813..c4b8b62 100644
--- a/examples/gpuExamples/cavity3d/cavity3d.cpp
+++ b/examples/gpuExamples/cavity3d/cavity3d.cpp
@@ -330,7 +330,7 @@ int main(int argc, char *argv[])
 
     pcout << "Number of MPI threads: " << global::mpi().getSize() << endl;
 
-    T maxIter = 2000;   // Total number of iterations
+    T maxIter = 200;   // Total number of iterations
     T vtkIter = 10000;  // Frequency of VTK file output in production mode
     T statIter = 1000;  // Frequency at which stats are outputed, in production mode
     T startAvgIter =
