--- external/petsc325/upstream/include/petscversion.h	2026-09-19 18:38:22.620148701 +0200
+++ external/petsc325/petsc-3.25.5/include/petscversion.h	2026-08-30 15:35:17.000000000 +0200
@@ -9,14 +9,14 @@
 #define PETSC_VERSION_MINOR      25
 #define PETSC_VERSION_SUBMINOR   5
 #define PETSC_RELEASE_DATE       "Mar 30, 2026"
-#define PETSC_VERSION_DATE       "unknown"
+#define PETSC_VERSION_DATE       "Aug 30, 2026"
 
 #if !defined(PETSC_VERSION_GIT)
-#define PETSC_VERSION_GIT        "unknown"
+#define PETSC_VERSION_GIT        "v3.25.5"
 #endif
 
 #if !defined(PETSC_VERSION_DATE_GIT)
-#define PETSC_VERSION_DATE_GIT   "unknown"
+#define PETSC_VERSION_DATE_GIT   "2026-08-30 07:59:06 -0500"
 #endif
 
 #define PETSC_VERSION_EQ(MAJOR,MINOR,SUBMINOR) \

--- external/petsc325/upstream/lib/petsc/conf/rules	2026-09-19 18:38:22.640148949 +0200
+++ external/petsc325/petsc-3.25.5/lib/petsc/conf/rules	2026-08-30 15:35:17.000000000 +0200
@@ -258,7 +258,6 @@
 	$(subst $@, $@.o, ${PETSC_COMPILE_SINGLE}) $<
 	${CLINKER} -s MAIN_MODULE -s ASSERTIONS=2 -s SAFE_HEAP=1 -s STACK_OVERFLOW_CHECK=1 -s ALLOW_MEMORY_GROWTH $(EXEFLAGS) -o $@.js $@.o $(PETSC_LIB)
 	-@printf '#!/usr/bin/env sh\nnode --redirect-warnings=/dev/null $@.js $$* | grep -v "Heap resize call from" ' > $@
-	-@chmod u+x $@
 endif
 
 help-applications:
